src/gui/qgscoordinatereferencesystemmodel.cpp

```diff
@@ -0,0 +1,751 @@
+              tr( "User Defined Coordinate Systems" ),
+              QgsApplication::getThemeIcon( QStringLiteral( "/user.svg" ) ),  QStringLiteral( "USER" ) );
+        beginInsertRows( QModelIndex(), mRootNode->children().length(), mRootNode->children().length() );
+        mRootNode->addChildNode( newGroup.get() );
+        endInsertRows();
+        group = newGroup.release();
+      }
+
+      const QModelIndex parentGroupIndex = node2index( group );
+
+      beginInsertRows( parentGroupIndex, group->children().size(), group->children().size() );
+      QgsCoordinateReferenceSystemModelCrsNode *crsNode = addRecord( userRecord );
+      crsNode->setProj( details.proj );
+      crsNode->setWkt( details.wkt );
+      endInsertRows();
+      break;
+    }
+  }
+}
+
+void QgsCoordinateReferenceSystemModel::userCrsRemoved( long id )
+{
+  QgsCoordinateReferenceSystemModelGroupNode *group = mRootNode->getChildGroupNode( QStringLiteral( "USER" ) );
+  if ( group )
+  {
+    for ( int row = 0; row < group->children().size(); ++row )
+    {
+      if ( QgsCoordinateReferenceSystemModelCrsNode *crsNode = dynamic_cast< QgsCoordinateReferenceSystemModelCrsNode * >( group->children().at( row ) ) )
+      {
+        if ( crsNode->record().authId == QString::number( id ) )
+        {
+          const QModelIndex parentIndex = node2index( group );
+          beginRemoveRows( parentIndex, row, row );
+          delete group->takeChild( crsNode );
+          endRemoveRows();
+          return;
+        }
+      }
+    }
+  }
+}
+
+void QgsCoordinateReferenceSystemModel::userCrsChanged( const QString &id )
+{
+  QgsCoordinateReferenceSystemModelGroupNode *group = mRootNode->getChildGroupNode( QStringLiteral( "USER" ) );
+  if ( group )
+  {
+    for ( int row = 0; row < group->children().size(); ++row )
+    {
+      if ( QgsCoordinateReferenceSystemModelCrsNode *crsNode = dynamic_cast< QgsCoordinateReferenceSystemModelCrsNode * >( group->children().at( row ) ) )
+      {
+        if ( QStringLiteral( "USER:%1" ).arg( crsNode->record().authId ) == id )
+        {
+          // treat a change as a remove + add operation
+          const QModelIndex parentIndex = node2index( group );
+          beginRemoveRows( parentIndex, row, row );
+          delete group->takeChild( crsNode );
+          endRemoveRows();
+
+          userCrsAdded( id );
+          return;
+        }
+      }
+    }
+  }
+}
+
+QgsCoordinateReferenceSystemModelCrsNode *QgsCoordinateReferenceSystemModel::addRecord( const QgsCrsDbRecord &record )
+{
+  QgsCoordinateReferenceSystemModelGroupNode *parentNode = mRootNode.get();
+  std::unique_ptr< QgsCoordinateReferenceSystemModelCrsNode > crsNode = std::make_unique< QgsCoordinateReferenceSystemModelCrsNode>( record );
+
+  QString groupName;
+  QString groupId;
+  QIcon groupIcon;
+  if ( record.authName == QLatin1String( "USER" ) )
+  {
+    groupName = tr( "User Defined Coordinate Systems" );
```

Comment:
User-defined or user defined? Also, should it be User-defined Coordinate >>Reference<< Systems?